Control Allocation Sequential Desaturation unit tests

somebody-once-told-me opened the pull request:
### Solved Problem
Unit tests for ControlAllocationSequentialDesaturation. No code changes are made to any existing library, this only adds unit tests. The unit tests serve 2 purposes: 1) help ensure that expected functionality remains the same and 2) documents and demonstrates (through examples) the desaturation algorithm.

somebody-once-told-me commented:
hmm, looks like the tailsitter SITL test is failing? curious as this only adds a unit test.

somebody-once-told-me commented:
merged; now all SITL tests are passing.

sfuhrer commented:
Thanks for your contribution, I'm going through it later.

>hmm, looks like the tailsitter SITL test is failing? curious as this only adds a unit test.

It's now passing. These SITL tests are a bit flaky. 

Can you rebase your branch on latest main? Something seems to have been messed up with your history, and this PR shows up many commits that don't come from you.

somebody-once-told-me commented:
@sfuhrer I can try, but I am not sure why those other commits are showing up. Although, the changes are accurately captured, as can be seen in the `Files changed` tab. If this PR is merged as-is, won't it all be squashed in a single commit anyway (and the weird commit history shown above won't matter)?

somebody-once-told-me commented:
@sfuhrer the commit history has been cleaned up.

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,383 @@
+	setup_quad_allocator(allocator);
+	matrix::Vector<float, ActuatorEffectiveness::NUM_AXES> control_sp;
+	// Negative, because +z is "downward".
+	constexpr float THRUST_Z_TOTAL{-0.75};
```
```suggestion
	constexpr float THRUST_Z_TOTAL{-0.75f};
```

sfuhrer replied:
I generally like being very explicit.

somebody-once-told-me replied:
is this something that can be flagged in `make format`?

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,383 @@
+	allocator.allocate();
+
+	const auto &actuator_sp = allocator.getActuatorSetpoint();
+	constexpr float THRUST_Z_PER_MOTOR{-THRUST_Z_TOTAL / 4};
```
```suggestion
	constexpr float THRUST_Z_PER_MOTOR{-THRUST_Z_TOTAL / 4.f};
```
Generally best practice to only use floats in divisions.

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,383 @@
+	const auto &actuator_sp = allocator.getActuatorSetpoint();
+	// This value is based off of the effectiveness matrix. If the effectiveness matrix is changed,
+	// this will need to be changed.
+	constexpr float YAW_DIFF_PER_MOTOR{YAW_CONTROL_SP * 5};
```
Could you maybe define the "5" as a constant by itself?

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,383 @@
+
+	const auto &actuator_sp = allocator.getActuatorSetpoint();
+	// In the case of yaw saturation, thrust per motor will be reduced by the hard-coded
+	// magic-number yaw margin of 0.15f.
```
@MaEtUgR maybe it's a good time to document this magic number in the algorithm as well!

MaEtUgR replied:
I can only imagine why this is useful. I think @bresch had the full context here.

sfuhrer reviewed (commented):
Thanks for kicking this off, that's very valuable. 
I only have some minor things. I leave it up to you to also add some test cases for the "normalized" allocation - could though also be a follow up PR.

somebody-once-told-me commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,383 @@
+}
+
+// Returns true if the 2 input values are within TOLERANCE of each other.
+bool is_similar(float a, float b)
```
good question; after looking into it, looks like `EXPECT_FLOAT_EQ` applies tolerance of 1e-4, so I will use that.

somebody-once-told-me replied:
apparently `EXPECT_FLOAT_EQ` applies a tighter tolerance. I was getting a failure for values 0.30000025 and 0.30000004. So, I am using `EXPECT_NEAR` with tolerance 1e-4.

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,385 @@
+	);
+}
+
+static constexpr float EXPECT_NEAR_TOL{1e-4};
```
```suggestion
static constexpr float EXPECT_NEAR_TOL{1e-4f};
```

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,385 @@
+	ControlAllocationSequentialDesaturation allocator;
+	setup_quad_allocator(allocator);
+	matrix::Vector<float, ActuatorEffectiveness::NUM_AXES> control_sp;
+	control_sp(ControlAllocation::ControlAxis::ROLL) = 0;
```
```suggestion
	control_sp(ControlAllocation::ControlAxis::ROLL) = 0.f;
```

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,385 @@
+	allocator.allocate();
+
+	const auto &actuator_sp = allocator.getActuatorSetpoint();
+	constexpr float MOTOR_COUNT{4.f};
```
```suggestion
	constexpr int MOTOR_COUNT{4};
```
As the motor count is a countable physical object. I know I previously advised to use /4.f, but if you pull it out as a variable like you do now (which is a good thing!) then I would not cover a countable physical object by a float but rather have float/integer in the math.

sfuhrer reviewed (commented):
Just some mini comments, otherwise good to go.

PS: please do not force push a branch that you opened already a pull request for, as otherwise it's not possible for the reviewer to see what has changed since the last review. Just add the review comments as new commits on top, and push the branch like that. Once the PR is approved you can clean up the commits or we just squash-merge to merge it all into one commit.

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,385 @@
+	constexpr float YAW_EFFECTIVENESS_FACTOR{5.f};
+	constexpr float YAW_DIFF_PER_MOTOR{YAW_CONTROL_SP * YAW_EFFECTIVENESS_FACTOR};
+	// At yaw condition, there will be 2 different actuator values.
+	constexpr float MOTOR_COUNT{4.f};
```
Same as above

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,385 @@
+
+	const auto &actuator_sp = allocator.getActuatorSetpoint();
+	// At max yaw, only 2 motors will carry all of the thrust.
+	constexpr float YAW_MOTORS{2.f};
```
Same as above

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,385 @@
+	const auto &actuator_sp = allocator.getActuatorSetpoint();
+	// This value is based off of the effectiveness matrix. If the effectiveness matrix is changed,
+	// this will need to be changed.
+	constexpr float MOTOR_COUNT{4.f};
```
Same as above

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,385 @@
+	matrix::Vector<float, ActuatorEffectiveness::NUM_AXES> control_sp;
+	// Negative, because +z is "downward".
+	constexpr float DESIRED_THRUST_Z_PER_MOTOR{0.8f};
+	constexpr float MOTOR_COUNT{4.f};
```
Same as above

sfuhrer commented on src/modules/control_allocator/ControlAllocation/ControlAllocationSequentialDesaturationTest.cpp:
```diff
@@ -0,0 +1,385 @@
+	allocator.allocate();
+
+	const auto &actuator_sp = allocator.getActuatorSetpoint();
+	constexpr float MOTOR_COUNT{4.f};
```
Same as above

sfuhrer reviewed (commented):
You forgot a couple of MOTOR_COUNTS. Lets do them all in ints and then we can get this in!

somebody-once-told-me commented:
@sfuhrer thanks, I fixed those and also added the float specifier suffix to all float zeros, replaced magic numbers with `MOTOR_COUNT` where applicable, and changed `YAW_MOTORS` to an int.

sfuhrer reviewed (approved):
Thanks for the last clean up, let's get it in!
